File: app/schemas/friend.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.friend import FriendRequestStatus
from app.models.user import Position


class FriendRead(BaseModel):
    """One row in the friends list -- same shape as team.TeamMemberRead
    minus is_captain (friendship has no roles)."""

    id: uuid.UUID
    first_name: str
    last_name: str
    avatar_url: str | None = None
    level: int
    jersey_number: int | None = None
    position: Position | None = None


class FriendCodePayload(BaseModel):
    code: str = Field(min_length=1)


class FriendRequestSentRead(BaseModel):
    """Response to POST /friends/requests -- the receiver's info (not the
    sender's, who already knows who they are) so the UI can confirm "request
    sent to <name>" even though the sender only typed in a code.
    """

    id: uuid.UUID
    status: FriendRequestStatus
    receiver_id: uuid.UUID
    receiver_first_name: str
    receiver_last_name: str
    receiver_avatar_url: str | None = None


class FriendRequestRead(BaseModel):
    """Assembled manually in FriendService (not from_attributes) -- a
    FriendRequest row plus the sender's display fields, same "flat, reused
    as-is" reasoning as team.TeamJoinRequestRead. Only the sender's info is
    carried since the only list this backs is "requests sent *to* me"
    (see FriendService.list_incoming_requests) -- the receiver is always
    the caller themself.
    """

    id: uuid.UUID
    sender_id: uuid.UUID
    sender_first_name: str
    sender_last_name: str
    sender_avatar_url: str | None = None
    status: FriendRequestStatus
    created_at: datetime

File: frontend/src/types/friend.ts
import type { Position } from './user'

export const FRIEND_REQUEST_STATUSES = ['pending', 'accepted', 'declined'] as const
export type FriendRequestStatus = (typeof FRIEND_REQUEST_STATUSES)[number]

// GET /friends -- one row per accepted friendship.
export interface FriendRead {
  id: string
  first_name: string
  last_name: string
  avatar_url: string | null
  level: number
  jersey_number: number | null
  position: Position | null
}

export interface FriendCodePayload {
  code: string
}

// Response to POST /friends/requests -- the receiver's info, since the
// sender only typed in a code and doesn't already know who that is.
export interface FriendRequestSentRead {
  id: string
  status: FriendRequestStatus
  receiver_id: string
  receiver_first_name: string
  receiver_last_name: string
  receiver_avatar_url: string | null
}

// GET /friends/requests -- requests sent *to* the caller.
export interface FriendRequestRead {
  id: string
  sender_id: string
  sender_first_name: string
  sender_last_name: string
  sender_avatar_url: string | null
  status: FriendRequestStatus
  created_at: string
}
